chatlearn/models/vllm_module_v2.py

```diff
@@ -0,0 +1,214 @@
+        # common method is '__init__'
+        assert len(common_methods) == 1
+        TorchModule.__init__(self, *args)
+        self.local_rank = 0
+        os.environ['LOCAL_RANK'] = '0'
+        if 'worker_module_name' in kwargs and 'worker_class_name' in kwargs:
+            RayWorkerWrapper.__init__(self, **kwargs) # pylint: disable=non-parent-init-called
+        os.environ['VLLM_HOST_IP'] = self.get_address()
+        self.engine = None
+
+    def setup(self):
+        """Set up model and load checkpoint"""
+        tokenizer = AutoTokenizer.from_pretrained(self.model_args['tokenizer'])
+        tokenizer.tokenizer = tokenizer
+        self.tokenizer = tokenizer
+
+    def setup_vllm(self, workers):
+        # setup vllm engine in rank 0
+        os.environ['VLLM_HOST_IP'] = self.get_address()
+        set_vllm_actors(workers)
+        parser = FlexibleArgumentParser()
+        parser = AsyncEngineArgs.add_cli_args(parser)
+        backup_sys_argv = sys.argv
+        dtype = self.model_args.get("dtype", "bfloat16")
+        if self.model_args.get("fp16", False):
+            dtype = "float16"
+        vllm_sys_argv = ["",
+                         f"--model={self.model_args['load']}",
+                         f"--tensor_parallel_size={self.module_args.tensor_model_parallel_size}",
+                         f"--pipeline_parallel_size={self.module_args.pipeline_model_parallel_size}",
+                         f"--dtype={dtype}",
+                         "--worker_use_ray",
+                         "--disable_custom_all_reduce"]
+        sys.argv = vllm_sys_argv
+        args = parser.parse_args()
+        engine_args = AsyncEngineArgs.from_cli_args(args)
+        self.engine = self.from_engine_args(engine_args)
+
+        sys.argv = backup_sys_argv
+        self.tokenizer = self.engine.engine.tokenizer
+
+    def from_engine_args(
+            self,
+            engine_args: AsyncEngineArgs,
+            engine_config: Optional[EngineConfig] = None,
+            start_engine_loop: bool = True,
+            usage_context: UsageContext = UsageContext.ENGINE_CONTEXT,
+            stat_loggers = None,
+    ) -> "AsyncLLMEngine":
+        """Creates an async LLM engine from the engine arguments."""
+        # Create the engine configs.
+
+        if engine_config is None:
+            engine_config = engine_args.create_engine_config()
+
+        executor_class = AsyncLLMEngine._get_executor_cls(engine_config)
+
+        # Create the async LLM engine.
+        engine = AsyncLLMEngine(
+            **engine_config.to_dict(),
+            executor_class=executor_class,
+            log_requests=not engine_args.disable_log_requests,
+            log_stats=not engine_args.disable_log_stats,
+            start_engine_loop=start_engine_loop,
+            usage_context=usage_context,
+            stat_loggers=stat_loggers,
+        )
+        return engine
+
+    async def generate_one_sample(self, prompt, sampling_param, request_id):
```

Comment:
Can we support generating multi results for the same prompt? maybe used in the scene like O1

Reply:
we need to design the API for multi results later